custom_components/localtuya/coordinator.py

```diff
@@ -428,40 +441,47 @@ async def set_dps(self, states):
     async def _async_refresh(self, _now):
         if self.connected:
             self.debug("Refreshing dps for device")
-            # This a workdaround for >= 3.4 devices, since there is an issue on waiting for the correct seqno
+            # This a workaround for >= 3.4 devices, since there is an issue on waiting for the correct seqno
             try:
                 await self._interface.update_dps(cid=self._node_id)
             except TimeoutError:
                 pass
 
+    async def _sleep(self, seconds) -> bool:
+        """Interruptable sleep. Returns True if cancelled."""
+        try:
+            await asyncio.sleep(seconds)
+        except asyncio.CancelledError:
+            self.debug(f"Sleep({seconds}) interrupted", force=True)
+            return True
+        return False
+
     async def _async_reconnect(self):
         """Task: continuously attempt to reconnect to the device."""
-        if self._reconnect_task:
-            return
-
-        self._reconnect_task = True
         attempts = 0
         while True:
             # for sub-devices, if it is reported as offline then no need for reconnect.
             if self.is_subdevice and self._subdevice_off_count >= MIN_OFFLINE_EVENTS:
-                await asyncio.sleep(1)
+                if await self._sleep(1):
```

Comment:
I don't understand this `_sleep` approach it's confused, asyncio sleep can be interrupt the issue happens to be before was related to new entity handler which was handled much better with your PR.

Reply:
I wrote above that I've made the tests that showed `asyncio.sleep` does not stop unless it is called in a `try` block.

Reply:
May I know how did you test that?, Yes it's true in some cases that it not always stop the task immediately however this is not in our case.

Reply:
When I implemented awaits, it stopped working with WiFi T&H battery powered sensor. I've added logging and found out that the tasks sleep. If asyncio.CancelledError was sent to them, log would contain "unhandled exception" records and it would not hang.

Reply:
This could happen if we actually awaited for the task, however assuming that we wait for them to clean up after canceling one close with can just use gather and return expectation to ignore cancel errors or use suppress. 

For the past days I was running LT with this PR with some changes, not with `_sleep` logic with low-power T&H, sub-device and parent devices and thinks works normally no issues with tasks handling at all, I abused reload and another tests running LT in more then 1 instance if device supports and things still works fine. It's not necessary that _reconnect task was the one raised "unhandled exception" however await after close could interrupt close function

here's a refactored version also note that cancel them togethers did resolve an issue w/ low-power device where it became unknown if reloaded twice quickly __also we don't need to await for shutdown entitles on close__

```py
    async def close(self):
        """Close connection and stop re-connect loop."""
        if self._is_closing:
            return

        self._is_closing = True

        tasks = [self._task_shutdown_entities, self._task_reconnect, self._task_connect]
        pending_tasks = [task for task in tasks if task and task.cancel()] 
        # Should we include close interface as well?
        # pending_tasks.append(self.abort_connect())
        # return_exceptions would ignore raising errors and return them instead.
        await asyncio.gather(*pending_tasks, return_exceptions=True)
        ## OR using suppress ##
        from contextlib import suppress # should be at the top
        with suppress(asyncio.CancelledError):
            await asyncio.gather(*pending_tasks)

        # Close subdevices first, to prevent them try to reconnect
        # after gateway disconnected.
        subdevices = list(self.sub_devices.values())
        for subdevice in subdevices:
            await subdevice.close()

        if self._unsub_new_entity is not None:
            self._unsub_new_entity()
            self._unsub_new_entity = None

        await self.abort_connect()
```

Now that we made "not_closing" check along with connected check then why not make connect property do that as well, __would it cause an issues__?

```py
    @property
    def connected(self):
        """Return if connected to device. NOTE: if _is_closing is true will return false."""
        return not self._is_closing and self._interface and self._interface.is_connected
```

Reply:
> why not make connect property do that as well, would it cause an issues?

There are many places with `if not self.connected` which won't work properly. Also, there are a few `if self.connected` that don't care about `_is_closing`. So, better explicitly check for `_is_closing ` when it is required than to create a problem to solve.

Reply:
I guess I went with the faster on cancellation, shutdown, reconnect aren't always running however even if it does it will cancel instantly unlike connect task however if somehow connect canceled before reconnect I thought there might be even a little chance it may run task connect again.

Reply:
Please hold on: I'm yet to analyse and test your main suggestion. The commit I made yesterday contains changes proven to be good.

Reply:
@xZetsubou I'm not experienced in Python like you. Could you suggest how to utilize `asyncio.gather()` here:
https://github.com/Lurker00/hass-localtuya/blob/31228396dde77f93d4d5dadffc1bc26d85b177ad/custom_components/localtuya/coordinator.py#L346-L350
?

Reply:
Instead we uses `asyncio.wait` usually gather used if we are waiting for results.

```py
        if close_subdevs := [
            asyncio.create_task(subdev.close()) for subdev in self.sub_devices.values()
        ]:
            await asyncio.wait(close_subdevs)
```

Reply:
I also don't think we should wait for sub-devices to close, just create_task is enough.

this works better for me, you may want to test it did made difference to me.
```py
        for subdev in self.sub_devices.values():
            asyncio.create_task(subdev.close())
```

Reply:
> this works better for me, you may want to test it did made difference to me

My system with 32 devices (3 gateways, 7 WiFi devices, 22 Zigbee and BLE sub-devices) shuts down in 0.3 seconds. How would I feel any difference?

Please note, that one of the goals is to ensure full graceful shutdown. Orphaned tasks is not what I want to leave.